Userland/Libraries/LibM/math.cpp

```diff
@@ -703,6 +707,141 @@ long double ceill(long double value) NOEXCEPT
     return internal_to_integer(value, RoundingMode::Up);
 }
 
+#else
+
+double round(double x) NOEXCEPT
+{
+    // Note: This is break-tie-away-from-zero, so not the hw's understaning of
+    //       "nearest", which would be towards even.
+    if (x == 0.)
+        return 0.;
+    if (x > 0)
+        return floor(x + .5);
+    return ceil(x - .5);
+}
+
+float roundf(float x) NOEXCEPT
+{
+    if (x == 0.f)
+        return 0.f;
+    if (x > 0)
+        return floorf(x + .5f);
+    return ceilf(x - .5f);
+}
+
+long double roundl(long double x) NOEXCEPT
+{
+    if (x == 0.L)
+        return 0.L;
+    if (x > 0)
+        return floorl(x + .5L);
+    return ceill(x - .5L);
+}
+
+long lroundf(float value) NOEXCEPT
+{
+    return static_cast<long>(roundf(value));
+}
+
+long lround(double value) NOEXCEPT
+{
+    return static_cast<long>(round(value));
+}
+
+long lroundl(long double value) NOEXCEPT
+{
+    return static_cast<long>(roundl(value));
+}
+
+long long llroundf(float value) NOEXCEPT
+{
+    return static_cast<long long>(roundf(value));
+}
+
+long long llround(double value) NOEXCEPT
+{
+    return static_cast<long long>(round(value));
+}
+
+long long llroundd(long double value) NOEXCEPT
+{
+    return static_cast<long long>(roundl(value));
+}
+
+float floorf(float value) NOEXCEPT
+{
+    auto cw = AK::get_cw_x87();
+    auto cw2 = cw;
+    cw2.rounding_control = AK::RoundingMode::DOWN;
+    AK::set_cw_x87(cw2);
+    asm("frndint"
+        : "+t"(value));
+    AK::set_cw_x87(cw);
+    return value;
+}
+
+double floor(double value) NOEXCEPT
+{
+    auto cw = AK::get_cw_x87();
+    auto cw2 = cw;
+    cw2.rounding_control = AK::RoundingMode::DOWN;
+    AK::set_cw_x87(cw2);
+    asm("frndint"
+        : "+t"(value));
+    AK::set_cw_x87(cw);
+    return value;
```

Comment:
Can we make an RAII helper for setting and resetting the x87 control word?
Something like...
```cpp
{
    X87RoundingModeScope scope(X87RoundingMode::Down);
    asm("frndint"
        : "+t"(value));
    return value;
}
```

Also, don't we need to specify clobbers when using inline assembly like this?

Reply:
To the scoping thing, sure can do.

To the clobbers:
We aren't clobbering anything that we don't have as an output, so I don't think so